### ISSUE TITLE
Ability to ignore .d.ts files except when declaration": true

### ISSUE THREAD

**Issue by rasmussyberg** [ef-carbon/tspm#26]
Could be nice to have the ability the ignore the declarations as .d.ts as long as declaration": true is not set in tsconfig.json

The tool will fail if .d.ts files are not moved to outDir before running because it is looking for the file if in  include

**Comment by mattyclarkson** [comment]
It **should** ignore them when `declaration` is `false` as checked [here](https://github.com/ef-carbon/tspm/blob/b0be24f09e5e042525b3e163fc3e552b057d2257/lib/Mapper.ts#L56)

**Comment by rasmussyberg** [comment]
Sorry dont know why i have the problem then. I have not been working with tspm the last two weeks, so i haven't checked what im doing wrong. I consider it closed.

**Comment by victordidenko** [comment]
I have same issue. I guess, `fileNames` should be filtered [here](https://github.com/ef-carbon/tspm/blob/b0be24f09e5e042525b3e163fc3e552b057d2257/lib/Mapper.ts#L55)
Something like this:
```js
yield* this.parsed.fileNames
  .filter(path => options.declaration || !path.endsWith('.d.ts'))
  .map(path => new EsFile({ path, options, config: this.parsed }));
```
@mattyclarkson can you check this, please?

**Comment by mattyclarkson** [comment]
@victordidenko I can't reproduce this. `tspm` is self-hosting, i.e. it runs `tspm` on it's own output. If I change `declaration` in the `tsconfig.json` to `false`, I get no error. Could you guide me to reproduce the error?

This is because the `parsed.fileNames` are the *input* files. The same input file is given to `EsFile` *and* `TsFile`. `EsFile` changes the extension to `.js` and `TsFile` changes the extension to `.d.ts`. Therefore, filtering the parsed filenames will do nothing as all parsed files names are the original `.ts` files.

I agree that that there *will* be an error if the `.d.ts` files are not in the output directory before `tspm` runs. However, if `tsc` runs before `tspm` then it should always create those files.

I need to understand your setup more to determine why this error is occurring.

**Comment by victordidenko** [comment]
If .d.ts files located inside baseUrl, `tsc` will see them automatically, but it will not produce corresponding files in `outDir`.
`tspm` got .d.ts files in `fileNames`, changes extension to js (thus it got .d.js filename) and cannot open it, with error
> File not found: '.../dist/typings.d.js'

>

**Comment by victordidenko** [comment]
I'm not importing .d.ts files anywhere, so this is not modules, they are just type definitions. Typescript sees and respects them, but it doesn't compile them to output directory.

**Comment by victordidenko** [comment]
I've created repo with minimal setup to reproduce issue:
https://github.com/victordidenko/tspm-dts-error

Try
> yarn
> yarn build

`tsc` will successfully output file without compilation error.
But `tspm` (in postbuild step) will fail with error
> File not found: '.../tspm-dts-error/dist/typings.d.js'

**Comment by mattyclarkson** [comment]
Gotcha, will have a look at probably merge your fix. Give me a few.

**Comment by mattyclarkson** [comment]
@victordidenko have repro'd and #31 corrects it when `declarations` is `false`. Have added a comment to make it work in both cases. Thank you so much for you explanations and help 😄

**Comment by victordidenko** [comment]
Thank you for quick response :)

**Comment by mattyclarkson** [comment]
Fixed in [2.2.5](https://github.com/ef-carbon/tspm/releases/tag/v2.2.5)
